sdk/metric/meter.go

```diff
@@ -406,20 +406,28 @@ func (p int64ObservProvider) registerCallbacks(inst *observable[int64], cBacks [
 	}
 }
 
-func (p int64ObservProvider) callback(i *observable[int64], f instrument.Int64Callback) func(context.Context) error {
-	inst := callbackObserver[int64]{i}
+func (p int64ObservProvider) callback(i int64Observable, f instrument.Int64Callback) func(context.Context) error {
+	inst := int64Observer{i}
 	return func(ctx context.Context) error { return f(ctx, inst) }
 }
 
+type int64Observer struct {
+	int64Observable
+}
+
+func (o int64Observer) Observe(val int64, attrs ...attribute.KeyValue) {
+	o.observe(val, attrs)
+}
+
```

Comment:
> couldn't this be a function on `int64Observable` or `*observable[N]` and then we can use them directly?

I don't think this can be replaced with a function. It needs to satisfy the the `Int64Oberver` interface, meaning it needs an `Observe` method.

Assuming we would want to add a method to either of these types, there remain issues why that would be undesirable.

First the `int64Observer` also cannot be merged with the `float64Obsever` into a single type. The embedded `instrument.Asyncronous`, `instrument.Float64Observer`, `instrument.Int64Observer` have colliding private methods that cannot be resolved. This means that using something like `*observable[N]` cannot be used to satisfy both of these.

Also, if we wanted to extend the `int64Observable` with a public `Observe` method it would mean that any `int64Observable`, even if returned as an interface type, could be cast to an observer. Meaning `int64Observable.(interface{ Observe(...) }).Observe(...)` would be compilable and valid code and leaves us at the starting point of these changes.

Reply:
> using something like *observable[N] cannot be used to satisfy both of these.
Adding Observe(N, ...attr) to the *observable[N] will make both the float64Observable and the int64Observale satisfy their respective interfaces. Example: https://github.com/MadVikingGod/opentelemetry-go/blob/a94bd93b931430a809c7dffde501b5e90a9edd8a/sdk/metric/instrument.go#L270

> Meaning int64Observable.(interface{ Observe(...) }).Observe(...) would be compilable and valid code and leaves us at the starting point of these changes.

I think there is a substantial difference between our original position, a function that is on the interface we return but don't want users to use, and a function that happens to be on the type underlying an interface that is returned.

That being said I am not holding too tightly to this position, just want to make sure it has been considered.

Reply:
> > using something like *observable[N] cannot be used to satisfy both of these.
> 
> Adding Observe(N, ...attr) to the *observable[N] will make both the float64Observable and the int64Observale satisfy their respective interfaces. Example: https://github.com/MadVikingGod/opentelemetry-go/blob/a94bd93b931430a809c7dffde501b5e90a9edd8a/sdk/metric/instrument.go#L270

Right, looks like I miss-typed. The `*observable[N]` can implement `instrument.Asyncronous`, `instrument.Float64Observer`, and `instrument.Int64Observer` simultaneously, but it cannot implement the `instrument.Asyncronous`, `instrument.Float64Observerable`, and `instrument.Int64Observerable` interfaces simultaneously. There would be a private method resolution conflict. This would mean `int64Observable` and `float64Observable` would still be need to wrap the `*observable[N]` with the typed `instrument.*Observable`.

> I think there is a substantial difference between our original position, a function that is on the interface we return but don't want users to use, and a function that happens to be on the type underlying an interface that is returned.

True.

> That being said I am not holding too tightly to this position, just want to make sure it has been considered.

Sure. If it's all the same, I'm gonna leave it as it is now. I like the explicit allowance of access to an observation being controlled by this package. It is something we can revisit in the future if need be though given the unexported nature of things.